Refactor: Extract mod `fs` as crate `atomic-file-install`

NobodyXu opened the pull request:
To speedup codegen of `binstalk` and also publish a new crate that could be reused.

NobodyXu commented:
Oops, I got a checksum error when using crates.io git-index:

```
ERROR Fatal error:
  × For crate cargo-binstall: Expected checksum
  │ `0fd7ac067793240ca84b9deca1509c75e44b2fb046bc04e0ecc94930834540c5`, actual
  │ checksum `2faa572a84166fb49d465956e1494e2199f04e36b43b98e83e7bc5242a5a6195`
error: Recipe `e2e-test` failed on line 236 with exit code 76
  ╰─▶ Expected checksum
      `0fd7ac067793240ca84b9deca1509c75e44b2fb046bc04e0ecc94930834540c5`,
      actual checksum
      `2faa572a84166fb49d465956e1494e2199f04e36b43b98e83e7bc5242a5a6195`
```

NobodyXu commented:
Turns out that it is just spurious failure...
No idea what happened in crates.io

NobodyXu commented:
@passcod If you are satisfied with the name, I will merge, publish it and invite you to the new crate since our CI does not have the permission to do that.

NobodyXu commented:
@ryankurte Since the new crate is in MIT/Apache instead of GPL, I think we need approval from you.

passcod commented:
I ghink i'd prefer something like atomic-file-install to avoid ambiguity with other types of installs.

also be good to have crate documentation

NobodyXu commented:
> I ghink i'd prefer something like atomic-file-install to avoid ambiguity with other types of installs.

Thanks, I've renamed it to `atomic-file-install`.

> also be good to have crate documentation

I've added crate-level doc and improved doc for functions.

NobodyXu commented:
@passcod I've invited you to the newly published `atomic-file-install` crates.io
